src/halmos/sevm.py

```diff
@@ -1946,18 +2119,38 @@ def create(
         # new account address
         if op == EVM.CREATE:
             new_addr = ex.new_address()
-        else:  # EVM.CREATE2
+        elif op == EVM.CREATE2:  # EVM.CREATE2
             if isinstance(create_hexcode, bytes):
                 create_hexcode = con(
                     int.from_bytes(create_hexcode, "big"), len(create_hexcode) * 8
                 )
             code_hash = ex.sha3_data(create_hexcode, create_hexcode.size() // 8)
             hash_data = simplify(Concat(con(0xFF, 8), caller, salt, code_hash))
             new_addr = uint160(ex.sha3_data(hash_data, 85))
+        else:
+            raise HalmosException(f"Unknown CREATE opcode: {op}")
+
+        message = Message(
+            target=new_addr,
+            caller=caller,
+            value=value,
+            data=create_hexcode,
```

Comment:
why is the calldata set to the creation code here?

according to the evm spec, calldata should be set to empty:
https://github.com/ethereum/execution-specs/blob/master/src/ethereum/shanghai/vm/instructions/system.py#L116-L117

if it's only for the tracing purpose, i think it is better to create a new field for its own, rather than repurposing the calldata field.  i know that there is the calldata() helper function, but given that the calldata is crucial for correctness, i'd like to avoid any potential mistakes caused by this discrepancy in the future.

Reply:
here is how I ended up this design. We need to represent either a regular message (has a `target` and calldata but no code) or a create message (technically has no `target` address and no calldata but has code)

a. we could represent them both in the same class using disjoint fields (ie. separate `code` and `calldata` fields)
b. we could represent them both in the same class with overloaded fields
c. we could represent them using separate classes like `MessageCall` and `CreateCall`

I went with (b.) because the semantic difference is pretty small and easily handled with a helper function to know if `message.data` corresponds to code or calldata. I similarly "hijacked" the target field because it's convenient to track the address being executed in that context.

I'm not a fan of option (a.) because the semantics of that union object is weird and you could end up in an inconsistent state (having both `code` and `calldata`), I think if we're going to refactor going with option (c.) would be cleaner:
- it would be very clear at creation time if you're making a `MessageCall` or `CreateCall`
- checking if something is a message call or create call is done simply via a type check rather than a helper function
- they could easily be rendered/printed/traced differently
- it's impossible to end up in an inconsistent state

Reply:
ok, i think we need more time to discuss this further.  let's keep this as it is for the scope of this pr.